README.md
## apache_extensions 

mod_perl and mod_proxy can work together to provide authentication mechanisms 
within apache.  

In our case, on the va-ctt project, we must integrate the Department 
of Veterans Affairs Single Sign On software with software such as Nexus and Jenkins
despite the fact that this software does not integrate out of the box. 
This 
guide will document how to accomplish this.



mod_perl:


mod_perl must be installed against the apache webserver:
```
yum -y install mod_perl
```

after the install you should have a perl.conf file, 
for example:
```
/etc/httpd/conf.d/perl.conf
```



To validate your install add the following to perl.conf:


```
Alias /perl/ /var/www/cgi-bin/
<Location /perl>
    SetHandler perl-script
    PerlResponseHandler ModPerl::Registry
    PerlOptions +ParseHeaders
    Options +ExecCGI
    Order deny,allow
    Allow from all
</Location>
```



place cris.pl in:
```
/var/www/cgi-bin/
```

cris.pl is dependent on Perl's CGI.pm module (note some versions of CPAN do not require and will not allows the 'install' keyword):
```
cpan install CGI
```
When hitting your url, perhaps:
```
https://myserver/perl/cris.pl
```
If you see a page that shows you the mod perl version and your header info,
you have succeeded.



Now to configure header validation. In httpd.conf include this line:
```
PerlRequire /etc/httpd/scripts/startup.pl
```
Obviously, startup.pl should be there. startup.pl is what tells mod_perl
 how to find your various custom modules for extending apache. 
This file modifies
Perl's include path appropriately so perl can find all referenced source code.

Now in httpd.conf add:
```
<Location />
  PerlFixupHandler Prisme::ValidateHeader
</Location>
```


*NOTE::  This header validator can be configured in Apache's \<Location\>,
\<Directory\> or \<Files\> section.  
This current config will only run the validator 
against the root path ('/').  More research is needed to enhance the config to run where needed.

A quick examination of startup.pl shows us that:
```
/etc/httpd/scripts/modules
```
is on the include path.  We need to ensure that the subdirectory 'Prisme' is present
and that it contains the perl module 'ValidatePrisme.pm'

ValidatePrisme.pm needs the following software installed, (keep in mind 'install' might not be required):
```
cpan install LWP
cpan install JSON::Parse
```

There is a file called constants.pl, you must add it in httpd.conf:
```
PerlRequire /etc/httpd/scripts/constants.pl
```

I put it right below startup.pl.  This file is engineered to be easy for a sys admin to modify.
It looks partly like this:
```
#System administrator have at the following:
#number of seconds we will cache a users roles for
$SECONDS_CACHE = 5*60; #every five minutes re-fetch.
$PRISME_ROLES_URL = 'https://vaauscttdbs80.aac.va.gov:8080/rails_prisme/roles/get_ssoi_roles.json';
#$PRISME_ROLES_URL = 'http://localhost:3000/roles/get_ssoi_roles.json';
$REQUIRED_ROLES = ['super_user','administrator']; #at least one of these role must be present to prevent the dreaded 'FORBIDDEN'!
$CGI_USER_NAME = 'id';
$JSON_ROLE_NAME_KEY = 'name';
$LOG_HEADERS=0; #0 is false 1 is true, logs at the info log level.
$ACCEPT_ALL_REQUESTS=0; #if 0 works as expected (role validation occurs), if 1 role validation is disabled
```

our mod perl module does quite a bit of logging.  To set log levels look for something
like this in the appropriate .conf file
```
# Use separate log files for the SSL virtual host; note that LogLevel
# is not inherited from httpd.conf.
ErrorLog logs/ssl_error_log
TransferLog logs/ssl_access_log
LogLevel warn
```

The initial implementation allows Claudio's and Cris' accounts access and denies all
others.

*NOTE:: logging occurs in /var/log/httpd*

mod_proxy 

######################

mod_proxy:

The objective is to forward/proxy the incoming request over ssl to its final destination.
In this initial implementation, we are proxying / (root) over to the destination server.
We have to nest the mod_proxy config in the ssl.conf file under httpd/conf.d/ssl.conf
In this initial implementation we aren't using VirtualHost(s) yet.

mod_proxy config:


# We must load the following modules:
LoadModule proxy_module modules/mod_proxy.so
LoadModule proxy_http_module modules/mod_proxy_http.so

# Add the following directives:
ProxyRequests Off
ProxyPreserveHost On
ProxyPass        / https://vaservername:4848
ProxyPassReverse / https://vaservername:4848

# Need to research Access Control for the version of httpd we are running in the VAservers:
# http://httpd.apache.org/docs/2.4/upgrading.html#run-time

...


ChangeLog.md
apache extensions changelog 

This changelog summarizes changes and fixes which are a part of each revision.  For more details on the fixes, refer tracking numbers 
where provided, and the git commit history.  Note that this repository contains system configuration files - not software that is 
deployed.

THIS IS THE DEVEOPMENT BRANCH

* 2017/09/12 - 5.1
	* All configuration changes required for release 5 - including NTRT, etc, and new ansible scripts to ease deployments.

* 2017/03/21 - 1.9
	* renamed:server_config.yml-PRE -> server_config.yml-PRE-R2
    * renamed:server_config.yml-PROD -> server_config.yml-PROD-R2
    * renamed:ssl.conf-PRE -> ssl.conf-PRE-R2
    * renamed:ssl.conf-PROD -> ssl.conf-PROD-R2
    * Production build for Release 3

* 2017/03/02 - 1.8
    * updates to server index pages
    * updates to cleanup multi-deploy config files

* 2017/01/26 - 1.7 
    * New files: index.html-DEV, index.html-SQA, index.html-TEST
    * Changes to: ssl.conf-SQA (newciphers), ssl.conf-TEST (newcyphersuites)

* 2017/01/19 - 1.6 
    * Changes to be committed:multi-deploy
    * New file:server_config.yml-SQA-multi-deploy
    * New file:ssl.conf-SQA-multi-deploy

* 2017/01/05 - 1.5: 
    * Changes to be committed:proxy-multiple instances
    * modified:ssl.conf-DEV
    * modified:server_config.yml-DEV
    
* 2016/11/13 - 1.4: 
    * Added missing yaml files

* 2016/12/02 - 1.3: 
    * Config updates
    
* 2016/11/17 - 1.2: 
    * Config updates

* 2016/11/03 - 1.1: 
    * See the GIT changelog for updates prior to this release.


perl/cgi-bin/cris.pl
#!/usr/bin/perl
# a little script to show the headers our environment receives.
# changing the name of this script is punishable by death.
use CGI;
use strict;
use warnings;

my $query = CGI->new;
print $query->header('text/html')."\n";
print "Cris says hi<br>\n";
print "mod perl is ".$ENV{"MOD_PERL"}."<br>\n";

my $text = "";

#- Getting the request method
   $text .= "Request method = ".$query->request_method()."\n";

#- Getting input data from the query string and from the data content
   $text .= "Names and values from param():\n";
  my  @names = $query->param();
   foreach my $name (@names) {
      $text .= "   $name = ".$query->param($name)."\n";
   }

#- Getting request headers
   $text .= "Names and values from http():\n";
   @names = $query->http();
   foreach  my $name (@names) {
      $text .= "   $name = ".$query->http($name)."\n";
   }

   print $query->header();
   print $query->start_html(-title=>'CGI-pm-Request-Info.pl');
   print $query->pre($text);
   print $query->end_html();


perl/modules/Prisme/ValidateHeader.pm
#must be in a directory named 'Prisme'
package Prisme::ValidateHeader;

use strict;
use warnings;
use Apache2::Const qw(FORBIDDEN OK);
use Apache2::Log;
use Apache2::RequestRec;
use APR::Table;
use Apache2::URI();
use APR::URI();
use Apache2::RequestIO ();
#require 'constants.pl';# done in http.conf
use LWP::UserAgent;
use URI;
my $ua = LWP::UserAgent->new( ssl_opts => { verify_hostname => 0 } );
use JSON;
use Time::HiRes qw/gettimeofday/;


#print "My lease will last for $CONST::SECONDS_CACHE seconds!\n";
my %cache_hash;
$cache_hash{'lease'} = { };
$cache_hash{'roles'} = { };
my $json_decoder_ring = JSON->new;
my $pid_dir = "/tmp/apache_extensions";
my $pid_file = "$pid_dir/$$.pid";
my $total_rest_calls = 0;
my $total_rest_cache_uses = 0;

#used in debugging for easy pid creation tracking to analyze what apache is doing
sub pid_file($$$$) {
    my $user_name = shift;
    my $rest_time = shift;
    my $rest_calls = shift;
    my $cache_uses = shift;
    mkdir $pid_dir unless (-e $pid_dir);
    my $time = localtime;
    open(my $fh, '>>', $pid_file);
    say $fh "$$ made a rest fetch at ($user_name -- $rest_time (seconds)): $time. rest_call_count=$rest_calls ; rest_cache_use=$cache_uses";
    close $fh;
}


sub allowed($$$) {
    my $roles = shift;
    my $context = shift;
    my $logger = shift;
    my @role_names = @$roles;

    my %union = my %isect = ();
    no warnings;
    my $context_roles = $$CONST::ROLES_BY_CONTEXT{$context};
    my $configured_roles = $context_roles eq '' ? $CONST::DEFAULT_ROLES : $context_roles;
    my $role_string = join(',', @$configured_roles);
    $logger->info("The required roles are $role_string for context $context");
    foreach my $e (map {lc} @role_names, map {lc} @$configured_roles) {
        $union{$e}++ && $isect{$e}++;
    }
    use warnings;
    my @isect = keys %isect;
    if (scalar @isect) {
        #we have at least one role we need.
        #send OK
        #print "All is good!!!\n";
        $logger->info("Returning OK!");
        return OK;
    }
    else {
        #we have none of the roles we need
        #send forbidden
        #warn "You are a hacker!\n";
        $logger->info("Returning FORBIDDEN!");
        return FORBIDDEN;
    }
}

sub rest_call($$$) {
    my $user_name = shift;
    my $context = shift;
    my $logger = shift;
    my $return_code;
    $logger->info("Rest call for $user_name");
    unless (defined $cache_hash{'lease'}->{$user_name}) {
        $cache_hash{'lease'}->{$user_name} = 0;
        #print $cache_hash{'lease'}->{$user_name} . "\n";
        my $last_check = $CACHE::cache_hash{'lease'}->{$user_name};
        $logger->info("Last check for user $user_name was at $last_check");
    }
    if (( time - $cache_hash{'lease'}->{$user_name} ) > $CONST::SECONDS_CACHE)
    {
        $logger->info( "Rerunning role fetch for user $user_name");

        no warnings;
        my $url = URI->new($CONST::PRISME_ROLES_URL);
        $url->query_form( $CONST::CGI_USER_NAME => $user_name );
        $logger->info("Prisme URL is: $url");    # so we can see it
        # Create a request
        my $req = HTTP::Request->new( GET => $url );

        # Pass request to the user agent and get a response back
        my $res = 0;
        my $roles = 0;
        my $content = 0;
        eval {
            #we will trap any errors during the fetch and parse.
            #parsing will fail if we cannot cannect to Prisme or
            #if we get invalid JSON.  We will log the error, and send a forbidden.
            my $rest_fetch_time = gettimeofday;
            $res = $ua->request($req);
            $rest_fetch_time = gettimeofday - $rest_fetch_time;
            $content = $res->content;
            $logger->info("The roles from prisme are: $content, fetch time was $rest_fetch_time seconds.");
            $roles = $json_decoder_ring->decode($content)->{'roles'};
            if ($CONST::GENERATE_PID_FILE) {
                $total_rest_calls++;
                pid_file($user_name, $rest_fetch_time, $total_rest_calls, $total_rest_cache_uses );
            }
        };
        unless ($roles) {
            $logger->error("Failed to get Prisme roles!");
            $logger->error("Prisme returned: $content") if $content;
            $logger->error("$@");
            $logger->error("$!");
            $logger->error("Returning forbidden due to failure!");
            return FORBIDDEN;    #return FORBIDDEN code
        }
        my @role_names = @$roles; #map { $_->{$CONST::JSON_ROLE_NAME_KEY} } @$roles;
        $cache_hash{'roles'}->{$user_name} = \@role_names;
        $return_code = allowed( \@role_names, $context, $logger );
        use warnings;
        my $current_time = time;
        $cache_hash{'lease'}->{$user_name} = time;
        $logger->info("Setting time for $user_name to $current_time");
    }
    else {
        $total_rest_cache_uses++ if ($CONST::GENERATE_PID_FILE);
        $logger->info("Using role cache for user $user_name");
        $return_code = allowed( $cache_hash{'roles'}->{$user_name}, $context, $logger );
    }
    $return_code;
}

sub apr_iterator($$) {
    my $headers = shift;
    my $logger = shift;
    while (my ($key, $value) = each %$headers) {
        $logger->info("PID: $$ DATA: $key => $value");    #goes to  /var/log/httpd/error_log
    }
    return 1;
}

sub handler {
    my $r = shift;
    my $context = (grep {$_ ne ''} split('/', $r->parsed_uri()->rpath))[0];
#    my $path_info = $r->path_info();
#    $r->log->info("path info is $path_info");
    my $request_type = (split('\.',
        (grep {$_ ne ''} split('/', $r->parsed_uri()->rpath))[- 1]))[- 1]; #something like 'js' or 'css'
    my $auto_accept = grep {$_ eq $request_type} @$CONST::AUTO_ACCEPT;
    #my $last_path = (grep {$_ ne ''} split('/', $r->parsed_uri()->rpath))[-1]; #something like batch.css
    #$r->log->info("request_type is $request_type");
    if ($CONST::LOG_HEADERS) {
        $r->log->info("PID: DATA: -----------------------------------------------------------------");
        #$r->headers_in()->do("apr_iterator");
        apr_iterator($r->headers_in(), $r->log);
        $r->log->info("PID: : -----------------------------------------------------------------");
    }

    if ($CONST::ACCEPT_ALL_REQUESTS) {
        $r->log->warn("In accept all requests mode! $$");
        return OK;
    }
    #my $accept = $r->headers_in->get('Accept');
    my $user = $r->headers_in->get('ADSAMACCOUNTNAME');
    #$r->log->info("Request start on pid $$: The accept header is $accept");
    if ($user) {
        $r->log->info("Request start on pid $$: The user for this request is $user");
        my $val = rest_call($user, $context, $r->log);
        my $roles = $cache_hash{'roles'}->{$user};
        my $role_string = join(',', @$roles);
        $r->headers_in->set('prisme-roles' => $role_string );
        #$r->headers_out->add('prisme.roles2'=> $role_string );
        #	$r->err_headers_in->add('prisme-roles2', $role_string);
        #$|++;
        #$r->rflush();# $r->rflush can't be called before the response phase if using PerlFixupHandler Prisme::ValidateHeader
        $r->log->info("Request end on pid $$: The user for this request is $user, the roles are $role_string, returning $val");
        return $val; #OK or FORBIDDEN
    }
    else {
        if ($auto_accept) {
            # we accept if the request is strictly an image request.
            #Accept => image/png,
            $r->log->info("Request end on pid $$: Just a(n) $request_type request OK.");
            return OK;
        } else {
            $r->log->info("Request end on pid $$: There is no user. Forbidden.");
            return FORBIDDEN;
        }
    }

}
#$r->log_error("--->request: log_error");
#$r->log("request: regular log");
#warn "DATA: -----------------------------------------------------------------";
#$r->headers_in()->do("apr_iterator");
#warn "DATA: -----------------------------------------------------------------";

#note, the HTTP_ portion should not be included.
#return OK unless $r->headers_in->EXISTS('SM_UNIVERSALID');#Shouldn't need this line in PerlFixupHandler lifecycle phase
# in a meeting with:
#Ray Shapouri
#Mohammad.Shapouri@va.gov
#We have been advised to switch to:
#HTTP_ADSAMACCOUNTNAME
#$r->log->info("CRIS 1--->request: log_info");
#$r->log_error("The user is $user");
#return OK if (($user eq 'vhaiswshuppc') || ($user eq 'VHAISPBOWMAG') ||($user eq 'vacocuestc'));
#return FORBIDDEN;

1;

perl/modules/Prisme/ValidatePrisme.pm
#must be in a directory named 'Prisme'
package Prisme::ValidateHeader;

use strict;
use warnings;
use Apache2::Const qw(FORBIDDEN OK);
use Apache2::Log;
use Apache2::RequestRec;
use APR::Table;
#require 'constants.pl';# done in http.conf
use LWP::UserAgent;
use URI;
my $ua = LWP::UserAgent->new( ssl_opts => { verify_hostname => 0 } );
use JSON;

#print "My lease will last for $CONST::SECONDS_CACHE seconds!\n";
my %cache_hash;
$cache_hash{'lease'} = {};
$cache_hash{'roles'} = {};
my $json_decoder_ring = JSON->new;

sub allowed($$) {
	my $roles      = shift;
	my $logger = shift;
	my @role_names = @$roles;

	my %union = my %isect = ();
	no warnings;
	foreach my $e ( @role_names, @$CONST::REQUIRED_ROLES ) {
		$union{$e}++ && $isect{$e}++;
	}
	use warnings;
	my @isect = keys %isect;
	if ( scalar @isect ) {
		#we have at least one role we need.
		#send OK
		#print "All is good!!!\n";
		$logger->info("Returning OK!");
		return OK;
	}
	else {
		#we have none of the roles we need
		#send forbidden
		#warn "You are a hacker!\n";
		$logger->info("Returning FORBIDDEN!");
		return FORBIDDEN;
	}
}

sub rest_call($$) {
	my $user_name = shift;
	my $logger = shift;
	my $return_code;
	$logger->info("Rest call for $user_name");
	unless ( defined $cache_hash{'lease'}->{$user_name} ) {
		$cache_hash{'lease'}->{$user_name} = 0;
		#print $cache_hash{'lease'}->{$user_name} . "\n";
		my $last_check = $CACHE::cache_hash{'lease'}->{$user_name};
		$logger->info("Last check for user $user_name was at $last_check");
	}
	if ( ( time - $cache_hash{'lease'}->{$user_name} ) > $CONST::SECONDS_CACHE )
	{
		$logger->info( "Rerunning role fetch for user $user_name");

		no warnings;
		my $url = URI->new($CONST::PRISME_ROLES_URL);
		$url->query_form( $CONST::CGI_USER_NAME => $user_name );
		$logger->info("Prisme URL is: $url");    # so we can see it
		                     # Create a request
		my $req = HTTP::Request->new( GET => $url );

		# Pass request to the user agent and get a response back
		my $res     = 0;
		my $roles   = 0;
		my $content = 0;
		eval {
		  #we will trap any errors during the fetch and parse.
		  #parsing will fail if we cannot cannect to Prisme or
		  #if we get invalid JSON.  We will log the error, and send a forbidden.
			$res     = $ua->request($req);
			$content = $res->content;
			$logger->info("The roles from prisme are: $content");
			$roles   = $json_decoder_ring->decode($content)->{'roles'};
		};
		unless ($roles) {
			$logger->error("Failed to get Prisme roles!");
			$logger->error("Prisme returned: $content") if $content;
			$logger->error("$@");
			$logger->error("$!");
			$logger->error("Returning forbidden due to failure!");
			return FORBIDDEN;    #return FORBIDDEN code
		}
		my @role_names =  @$roles; #map { $_->{$CONST::JSON_ROLE_NAME_KEY} } @$roles;
		$cache_hash{'roles'}->{$user_name} = \@role_names;
		$return_code = allowed( \@role_names,$logger );
		use warnings;
		my $current_time = time;
		$cache_hash{'lease'}->{$user_name} = time;
        $logger->info("Setting time for $user_name to $current_time");
	}
	else {
		$logger->info("Using role cache for user $user_name");
		$return_code = allowed( $cache_hash{'roles'}->{$user_name}, $logger );
	}
	$return_code;
}

sub apr_iterator($$) {
    my $headers = shift;
    my $logger = shift;
	while (my ($key, $value) = each %$headers) {
		$logger->info("PID: $$ DATA: $key => $value");    #goes to  /var/log/httpd/error_log
	}
	return 1;
}

sub handler {
	my $r = shift;
	if ($CONST::LOG_HEADERS) {
    	$r->log->info("PID: DATA: -----------------------------------------------------------------");
        #$r->headers_in()->do("apr_iterator");
        apr_iterator($r->headers_in(), $r->log);
        $r->log->info("PID: : -----------------------------------------------------------------");
    }

	if ($CONST::ACCEPT_ALL_REQUESTS) {
	  $r->log->warn("In accept all requests mode! $$");
	  return OK;
	}
	my $accept = $r->headers_in->get('Accept');
	my $user = $r->headers_in->get('ADSAMACCOUNTNAME');
	$r->log->info("Request start on pid $$: The user for this request is $user");

	if($user) {
		my $val = rest_call($user,$r->log);
    	$r->log->info("Request end on pid $$: The user for this request is $user");
    	return $val; #OK or FORBIDDEN
    	}
    	else {
    	     if ($accept =~ /image\/png/) {
                  # we accept if the request is strictly an image request.
                  #Accept => image/png,
                  $r->log->info("Request end on pid $$: Just an image request OK.");
                  return OK;
            } else {
    	          return FORBIDDEN;
    	      	  $r->log->info("Request end on pid $$: There is no user. Forbidden.");
    	      	}
	}

}
	#$r->log_error("--->request: log_error");
	#$r->log("request: regular log");
	#warn "DATA: -----------------------------------------------------------------";
	#$r->headers_in()->do("apr_iterator");
	#warn "DATA: -----------------------------------------------------------------";

#note, the HTTP_ portion should not be included.
#return OK unless $r->headers_in->EXISTS('SM_UNIVERSALID');#Shouldn't need this line in PerlFixupHandler lifecycle phase
# in a meeting with:
#Ray Shapouri
#Mohammad.Shapouri@va.gov
#We have been advised to switch to:
#HTTP_ADSAMACCOUNTNAME
	#$r->log->info("CRIS 1--->request: log_info");
	#$r->log_error("The user is $user");
	#return OK if (($user eq 'vhaiswshuppc') || ($user eq 'VHAISPBOWMAG') ||($user eq 'vacocuestc'));
	#return FORBIDDEN;

1;

perl/constants.pl
package CONST;
no warnings;

#System administrator have at the following:
#number of seconds we will cache a users roles for
$SECONDS_CACHE = 5*60; #every five minutes re-fetch.
$PRISME_ROLES_URL = 'https://vaauscttdbs80.aac.va.gov:8080/rails_prisme/roles/get_ssoi_roles.json';
#$PRISME_ROLES_URL = 'http://localhost:3000/roles/get_ssoi_roles.json';
$ROLES_BY_CONTEXT = {'ntrt' => ['super_user', 'ntrt_user', 'ntrt_staff', 'ntrt_admin']}; #roles tied to the context
#$ROLES_BY_CONTEXT = {'jenkins' => ['super_user', 'Approver'], 'ntrt' => ['super_user', 'NTRT']}; #multi context example
$DEFAULT_ROLES = ['super_user','administrator']; # If the context is not listed above then at least one of these role must be present to prevent the dreaded 'FORBIDDEN'!
$CGI_USER_NAME = 'id';
$LOG_HEADERS=0; #0 is false 1 is true, logs at the info log level.
$ACCEPT_ALL_REQUESTS=0; #if 0 works as expected (role validation occurs), if 1 role validation is disabled
$AUTO_ACCEPT=['css','js','png','image','jpg','jpeg','class','gif','fcc','scc','sfcc','ccc','ntc', 'ico','map','eot'];#requests of this type will be auto accepted (they will never be forbidden), so http://www.foo.com/faa/fi.js would be accepted.
1;